src/jit/codegenxarch.cpp

```diff
@@ -7423,29 +7426,237 @@ unsigned CodeGen::getBaseVarForPutArgStk(GenTreePtr treeNode)
+        assert(!varTypeIsLong(fieldType));
+        assert(fieldOffset <= prevFieldOffset);
+
+        // If the field is slot-like, we can use a push instruction to store the entire register no matter the type.
+        //
+        // The GC encoder requires that the stack remain 4-byte aligned at all times. Round the adjustment up
+        // to the next multiple of 4. If we are going to generate a `push` instruction, the adjustment must
+        // not require rounding.
+        // NOTE: if the field is of GC type, we must use a push instruction, since the emitter is not otherwise
+        // able to detect stores into the outgoing argument area of the stack on x86.
+        const bool fieldIsSlot = ((fieldOffset % 4) == 0) && ((prevFieldOffset - fieldOffset) >= 4);
+        int        adjustment  = roundUp(currentOffset - fieldOffset, 4);
+        if (fieldIsSlot)
+        {
+            fieldType         = genActualType(fieldType);
+            unsigned pushSize = genTypeSize(fieldType);
+            assert((pushSize % 4) == 0);
+            adjustment -= pushSize;
+            while (adjustment != 0)
+            {
+                inst_IV(INS_push, 0);
+                currentOffset -= pushSize;
+                genStackLevel += pushSize;
+                adjustment -= pushSize;
+            }
+            m_pushStkArg = true;
+        }
+        else
+        {
+            m_pushStkArg = false;
+            if (adjustment != 0)
+            {
+                // This moves the stack pointer to (fieldOffset + sizeof(fieldType)) to account for the `push`.
+                // For this case, we must adjust the stack and generate stack-relative stores rather than pushes.
+                // Adjust the stack pointer to the next slot boundary.
+                inst_RV_IV(INS_sub, REG_SPBASE, adjustment, EA_PTRSIZE);
+                currentOffset -= adjustment;
+                genStackLevel += adjustment;
+            }
+        }
+
+        // Consume the register, if any, for this field. Note that genConsumeRegs() will appropriately
+        // update the liveness info for a lclVar that has been marked RegOptional, which hasn't been
+        // assigned a register, and which is therefore contained.
+        // Unlike genConsumeReg(), it handles the case where no registers are being consumed.
+        genConsumeRegs(fieldNode);
+        regNumber argReg = fieldNode->isContainedSpillTemp() ? REG_NA : fieldNode->gtRegNum;
+
+        // If we can't push this field, it needs to be in a register so that we can store
+        // it to the stack location.
+        if (!m_pushStkArg)
+        {
+            bool needByteReg =
+                (varTypeIsByte(fieldType) && ((argReg == REG_NA) || ((genRegMask(argReg) & RBM_BYTE_REGS) == 0)));
+            if ((argReg == REG_NA) || needByteReg)
+            {
+                assert(varTypeIsIntegralOrI(fieldNode));
+                if (tmpReg == REG_NA)
```

Comment:
IMO we ought to move the initialization of `tmpReg` out of the loop. We could simply have something like:
```
regNumber tmpReg = REG_NA;
if (putArgStk->gtRsvdRegs != RBM_NONE)
{
    assert(genCountBits(putArgStk->gtRsvdRegs) == 1);
    tmpReg = genRegNumFromMask(putArgStk->gtRsvdRegs);
    assert(genIsValidIntReg(tmpReg));
}
```